WordPress/gutenberg
# Global styles: add revisions UI

Related to:

- https://github.com/WordPress/gutenberg/issues/49601 (parent issue)
- https://github.com/WordPress/gutenberg/issues/41232
- https://github.com/WordPress/gutenberg/issues/33094


## What?
This PR displays revisions to the global styles in the global styles sidebar (revisions for the wp_global_styles custom post type) in a timeline-like list.

At the moment is a limit imposed by the rest controller of 100 revisions.

The revisions panel will only be available once there are more than two revisions. This is because the theme global styles default state is not treated as a revision. It's [possible to add it in though](https://github.com/WordPress/gutenberg/pull/50089#issuecomment-1532324313). 

## How?
Building on the work in :

- https://github.com/WordPress/gutenberg/pull/49973
- https://github.com/WordPress/gutenberg/pull/49974

## TODO
- [x] Maybe extract components in screen-revisions.js into separate files
- [x] Extract changes that are testable on their own, e.g., [API changes](https://github.com/WordPress/gutenberg/pull/50234)
- [x] Whatever else swings by

## Testing Instructions
1. Make a few changes to your site's global styles, save each one.
2. Open the revisions view via the global styles context menu
3. Check that the "currently-saved" revision is correct. Now select a revision and ensure you can preview it
4. Click on the "Load revision" to load it in the editor and close the revisions view
5. Save the post
6. Make some changes to global styles but don't save. Changes may include global style properties such as color, or layout and also block-level styles or custom color palettes. The more the better. Go for it!
7. Open the revisions view and check that there is an "Unsaved changes" item at the start of the list. Now try to load a revision as per steps 2-4
8. You should be prompted to exit and save your current changes from step 6, or go ahead and load the revision into the editor
9. Now open the revisions view
10. Play around with the panel and make sure it closes as expected, e.g., when switching to code editor view or opening the stylebook

Run the E2E tests if you want
```
npm run test:e2e:playwright -- --headed site-editor/user-global-styles-revisions.spec
```

## Screenshots or screencast


https://user-images.githubusercontent.com/6458278/236083659-ec8e1e53-dc90-43d0-9701-7d5d07269965.mp4
